core/drivers/bcm_gpio.c

```diff
@@ -0,0 +1,148 @@
+
+#define IPROC_GPIO_REG(pin, reg) (GPIO_BANK(pin) * GPIO_BANK_SIZE + (reg))
+#define IPROC_GPIO_SHIFT(pin)    ((pin) % NGPIOS_PER_BANK)
+
+static inline void iproc_set_bit(struct gpio_chip *gc, unsigned int reg,
+				 unsigned int gpio)
+{
+	unsigned int offset = IPROC_GPIO_REG(gpio, reg);
+	unsigned int shift = IPROC_GPIO_SHIFT(gpio);
+
+	io_setbits32((vaddr_t)gc->base + offset, BIT(shift));
+}
+
+static inline void iproc_clr_bit(struct gpio_chip *gc, unsigned int reg,
+				 unsigned int gpio)
+{
+	unsigned int offset = IPROC_GPIO_REG(gpio, reg);
+	unsigned int shift = IPROC_GPIO_SHIFT(gpio);
+
+	io_clrbits32((vaddr_t)gc->base + offset, BIT(shift));
+}
+
+static void iproc_gpio_set(struct gpio_chip *gc, unsigned int gpio,
+			   enum gpio_level val)
+{
+	if (val == GPIO_LEVEL_HIGH)
+		iproc_set_bit(gc, IPROC_GPIO_DATA_OUT_OFFSET, gpio);
+	else
+		iproc_clr_bit(gc, IPROC_GPIO_DATA_OUT_OFFSET, gpio);
+}
+
+static enum gpio_level iproc_gpio_get(struct gpio_chip *gc, unsigned int gpio)
+{
+	unsigned int offset = IPROC_GPIO_REG(gpio, IPROC_GPIO_DATA_IN_OFFSET);
+	unsigned int shift = IPROC_GPIO_SHIFT(gpio);
+	int val;
+
+	val = io_read32((vaddr_t)gc->base + offset) & BIT(shift);
+
+	return val ? GPIO_LEVEL_HIGH : GPIO_LEVEL_LOW;
+}
+
+static void iproc_gpio_set_dir(struct gpio_chip *gc, unsigned int gpio,
+			      enum gpio_dir dir)
+{
+	if (dir == GPIO_DIR_OUT)
+		iproc_set_bit(gc, IPROC_GPIO_OUT_EN_OFFSET, gpio);
+	else
+		iproc_clr_bit(gc, IPROC_GPIO_OUT_EN_OFFSET, gpio);
+
+}
+
+static enum gpio_dir iproc_gpio_get_dir(struct gpio_chip *gc,
+					unsigned int gpio)
+{
+	unsigned int offset = IPROC_GPIO_REG(gpio, IPROC_GPIO_OUT_EN_OFFSET);
+	unsigned int shift = IPROC_GPIO_SHIFT(gpio);
+	uint32_t val;
+
+	val = io_read32((vaddr_t)gc->base + offset) & BIT(shift);
+
+	return val ? GPIO_DIR_OUT : GPIO_DIR_IN;
+}
+
+static enum gpio_interrupt iproc_gpio_get_itr(struct gpio_chip *gc,
+					  unsigned int gpio)
+{
+	unsigned int offset = IPROC_GPIO_REG(gpio, IPROC_GPIO_INT_MSK_OFFSET);
+	unsigned int shift = IPROC_GPIO_SHIFT(gpio);
+	uint32_t val;
+
+	val = io_read32((vaddr_t)gc->base + offset) & BIT(shift);
+
+	return val ? GPIO_INTERRUPT_ENABLE : GPIO_INTERRUPT_DISABLE;
+}
+
+static void iproc_gpio_set_itr(struct gpio_chip *gc, unsigned int gpio,
+			  enum gpio_interrupt ena_dis)
+{
+	if (ena_dis == GPIO_INTERRUPT_ENABLE)
+		iproc_set_bit(gc, IPROC_GPIO_OUT_EN_OFFSET, gpio);
+	else
+		iproc_clr_bit(gc, IPROC_GPIO_OUT_EN_OFFSET, gpio);
+
+}
+
+static const struct gpio_ops bcm_gpio_ops = {
+	.get_direction = iproc_gpio_get_dir,
+	.set_direction = iproc_gpio_set_dir,
+	.get_value = iproc_gpio_get,
+	.set_value = iproc_gpio_set,
+	.get_interrupt = iproc_gpio_get_itr,
+	.set_interrupt = iproc_gpio_set_itr,
+};
+KEEP_PAGER(bcm_gpio_ops);
+
+static void iproc_gpio_init(struct gpio_chip *gc, unsigned int paddr,
+			    unsigned int gpio_base, unsigned int ngpios)
+{
+	gc->base = (vaddr_t)phys_to_virt(paddr, MEM_AREA_IO_SEC);
```

Comment:
safer to call `core_mmu_get_va()` maybe, and also then cast is not required

Reply:
sure.